### ISSUE TITLE
Zelda Subscreen Delay REAL Fix

### ISSUE THREAD

**Issue by Fanatic-64** [project64/project64#20]
The pause menu in Zelda OoT takes several seconds to come up, and sometimes it will just freeze the game. Counter Factor=1 mitigates this to some degree. There is a GameShark code in the RDB which "fixes" this issue, but it will break the GLideN64 plugin's framebuffer emulation: https://gliden64.blogspot.com/2013/11/frame-buffer-emulation-part-i.html

Thus, we need a proper fix in the emulator so as to at least prevent the freezing.

**Comment by oddMLan** [comment]
I second that. The same thing happens in Paper Mario 64 when you open the subscreen inventory. 
In Paper Mario 64, when you press Start you get a inventory subscreen. In every emulator/gfx plugin combination I've tried, there is a delay of approximately 2 seconds between the sound cue when you press the Start button and when the screen shows up. The delay isn't actually meant to be there, it just happens when using an emulator. 

There's also a gameshark "fix" that completely breaks the framebuffer background in any plugin.

Interestingly, the ONLY plugin that correctly emulates these effects without ANY delay, is Angrylion's RDP, that doesn't have the delay but it is very resource-intensive, or a workaround would be using 1964 UltraFast with 4x Overclock, but it doesn't erradicate the problem itself, it just makes the delay shorter. The Wii virtual console version doesn't suffer from this problem.

Another observation, If I increase the counter factor in the emulator options, it acts as a multiplier of the delay.

In games like Zelda OOT and Paper Mario, Counter Factors acts like a multiplier of the delay.
If Counter Factor is set to 1, the subscreen delay is approximately 2 secs.
If Counter Factor is set to 2, the subscreen delay is approximately 4-5 secs.
If Counter Factor is set to 4, the subscreen delay is approximately 8-9 secs.

AFAIK, Angrylion's plugin bypasses some of the emulator's timing. Maybe that's why it is not affected.
Maybe someone could ask him what is he doing differently?

**Comment by cxd4** [comment]
While this issue is open, I guess I should mention that the GameShark cheat built into PJ64 for _Pokemon Snap_ is useless too with angrylion's RDP...perfectly playable with a complete graphics plugin, so the cheat just gets in the way to risk side effects and bugs like the cheats mentioned in this report.

**Comment by project64** [comment]
maybe I should have a setting to disable cheats in the rdb, so we can test easier then modifying the rdb.

In a video of CEN it looked like there was no pause there, so I would like to find a real solution to the pause.

**Comment by theboy181** [comment]
I need to mention that Pilot Wings 64 also has a cheat added for the "shadow fix" that is not needed for angrylion's plugin, and when its enabled it causes the game to crash after the first game play intro. This can be removed from the RDB and the problem is solved.

**Comment by project64** [comment]
was looking at the code to add an option to disable this .. turns out there already is an option to disable it. 

menu -> debugger -> r4300i -> Disable Game Fixes

**Comment by project64** [comment]
ok looking at this some more I tried lle gfx, lle audio could not boot zelda at all.

I tried your/angrylion rdp (http://forum.pj64-emu.com/showthread.php?t=4422) plugin with the disable game fixes disabled and it seems like it does not have the delay with the start screen.

**Comment by oddMLan** [comment]
Yeah, Angrylion plugin doesn't have the delay for some reason. I heard it bypasses some emulator timings but I'm not completely sure. Probably Angrylion would know more about the matter (does anybody know if he has an account here?)

**Comment by cxd4** [comment]
I can't remember if I've seen angrylion talk on GitHub GlideN64 yet...maybe he might have an account here.

But he does all his RDP development on GoogleCode, not GitHub:
https://code.google.com/p/angrylions-stuff/

That was how I got his e-mail address and occasionally sent him e-mails now and then asking him about VI/RDP technical questions to help me with changes.

**Comment by cxd4** [comment]
Pokemon Snap isn't the only thing that works on angrylion-RDP either.

When you go into Banjo-Tooie's, Super Smash Bros.'s "Screen Alignment" or "Screen Adjust" options, where it lets you re-center the position of the game screen...those also never worked in any HLE plugins.

I don't think he is doing anything that complicated in the RDP commands for timing.  I think it's more a matter of how the DRAM writeback/read conditions are timed while sorting out the triangle data, but I never thought of isolating exactly why that thing is supported.  I think I will ask him.

**Comment by cxd4** [comment]
On an unrelated note, @Fanatic-64, is this fine with you?
You can use a LLE plugin like angrylion's and use "Disable Game Fixes" in Project64 debugger?  If so I think there is not much left to keep this issue open for.

**Comment by cxd4** [comment]
I have just tested with angrylion's RDP.  It does not matter, whether "Disable Game Fixes" is on or off.  The subscreen load time is the same either way.  It's only slow if I switch to a different plugin like z64gl, and disable game fixes is on.

>ok looking at this some more I tried lle gfx, lle audio could not boot zelda at all.

I was able to boot Zelda just fine when I did this test and confirmed it was fixed in angrylion.
@Fanatic-64 what about you?  Isn't the subscreen fixed, with "Disable Game Fixes" turned on, and angrylion RDP?

**Comment by cxd4** [comment]
This issue really has nothing to do with Project64 itself as a core.  I've confirmed that by forcing the frame buffer in DRAM with black pixels, Zelda sub-screen has the delay in angrylion's plugin as well.  Conversely, if I fill it with white pixels, the sub-screen opens instantly with no delay in response.  So the issue isn't necessarily timing; it's frame buffer readback from RDRAM.  If the game's R4300-side CPU code is not happy with the results of the pixels stored into RDRAM (easily the case with a non-pixel-accurate graphics plugin), it is free to decline and fail the frame buffer read test--in which case, Zelda subscreen can fail, _Pokemon Snap_ cameras and motion events can fail, _Banjo-Tooie_ "screen alignment" options adjust can fail...all those FB read effects can fail.  Especially on z64gl, which does no frame buffer storage to `RDRAM + VI_ORIGIN_REG` whatsoever, and fails most demos, independent of HLE or LLE.

Jabo's "Copy Framebuffer to RDRAM" has it partially supported for _Pokemon Snap_, but the pixel buffer results are still volatile and can fail on occasion with "Disable Game Fixes" checked.  It doesn't seem to be accurate enough to write the correct pixel(s) to RDRAM for Zelda sub-screen though.  Perhaps the coverage information when determining the alpha bit is wrong on some edges, or perhaps while rendering primitives to fragments it does not copy to a separate frame buffer in render spans COPY cycle mode--one not seen traditionally.

In short, if we use a third-party graphics plugin, issue is non-issue.

**Comment by project64** [comment]
interesting so it is waiting till the rdp has finished drawing and determining that based on what is in the RDRAM. 

Maybe i need someway to link the cheat to jabo's plugin so I can disable it for other plugins

**Comment by cxd4** [comment]
Whatever the cheat for Pokemon Snap does, it only seems to have helped with Jabo's 1.6+ plugins, not 1.5.2 etc or any of RiceVideo/glN64/Glide64, certainly not z64gl.

Glide64 most likely can't store the information using the VI registers correctly--Gonetz has already discussed this in one of the blogs somebody linked.  He's planning to do more accurate VI emulation in GLideN64, but that invariably requires working with graphics on the software level.  You have to do several non-hardware-accelerated tasks, even if you use DirectX or OpenGL functions to do them for you often times they end up being done on the CPU side not on the GPU side.  So using the API functions can be even slower than manual C code.

Basically when an OpenGL plugin renders all the geometric primitives to a 320x240, 640x480, 1024x768 ... whatever size viewport the Project64 window screen is sized at, this is the native GL frame buffer on our own video card, but the RCP's VI registers dictate that only 237 scanlines, each 320 pixels wide, make up the active video output for the current part of the game.  So Jabo's Direct3D may be able to take the current client-side screen size, scale it down (possibly up) to the real N64 native resolution's target frame buffer size, then write the pixels to RDRAM.  Even then, however, the pixels have to be converted from standard RGB to 16-bit R5G5B5 in most cases, which actually is sometimes faster when done in pure C code rather than an OpenGL function like glReadPixels.

I was trying to test something like this in glReadPixels to PJGlide64 just now by putting it at the end of `ProcessDList` but wasn't successful at it.  I guess when the triangles are drawn the origin reg may have changed, or the scanlines can be subtly offset by the other VI regs, which also are unemulated by all other plugins as of yet.

**Comment by project64** [comment]
I am thinking I might modify the rdb handler of the cheats to be able to specify that it work just with certain plugins.

something like "gfx: plugin name", or maybe just plugin name and if any of the plugin types match then accept it. 

if you want it to go to multiple plugins then you need to have the entry in for each plugin that it is needed for. 

that way the cheat can stay for jabo's plugins, but not for the new gliden64 plugin

**Comment by Fanatic-64** [comment]
Maybe something like Cheat0(Jabo)=XXXX XXXXXXXX, also with an option to specify (Core) for cheats that should remain activated for all plugins (like Chamelon Twist Mempack Fix).

**Comment by cxd4** [comment]
As far as I know every plugin except angrylion and my version of it seem to have the sub-screen delay, so it would need to be added a lot of times for the other plugins besides Jabo in case other people wanted to test without knowing about the hack bypass situation with the cheats method.

If it works in GLideN64, then it may be better to just default to "Disable Game Fixes" being turned on, not off, rather than worrying about global hacks for all plugins or individual ones, and probably have GLideN64 as default plugin.

**Comment by project64** [comment]
if and when GLideN64 is the default plugin then we can see if I can do that, I am more looking at in general how do I make pj64 work correct out of the box. Not looking so much at helping every gfx plugin. Especially as Jabo's plugins are no longer developed or can be changed, mostly just looking at how I help fix that with out blocking future plugins.

would look like:
Core Note=(see GameFAQ)
Cheat0=801DB78B 0002 //Subscreen Delay Fix
Cheat0_plugin=Jabo 1.7 dx
Cheat1=D109A814 0320,8109A814 0000,D109A816 F809,8109A816 0000 //End Credits Fix
Cheat1_plugin=Jabo 1.7 dx
Clear Frame=0

**Comment by cxd4** [comment]
I guess it depends exactly what GLideN64 does and does not fix.  It will possibly fix Pokemon Snap and support VI-accurate FB writes, but I haven't figured out the algorithm yet behind how the Zelda sub-screen delay pause thing works.  It seems as if it's trying to search for a white pixel to determine the reset phase of the RDP to draw on a new canvas, and logs the frame buffer over multiple writes.  You could potentially be thinking too early about investing this extra work of implementation for extra RDB analysis in the emu, when we don't know yet exactly how much as well as how little GLideN64 will fix.

Hypothetically though, if GLideN64 fixes all the FB communications issues that these built-in cheats address, I would say just don't bother with the RDB, just make "Enable Game Fixes" only be effective when Jabo's 1.6/1.7 plugin is used regardless of the game or RDB.  Much easier to implement and saves more time. I think that would be at least as successful at achieving your preference to avoid "blocking out future plugins", as making it game-specific within the RDB and just saying Jabo's or Jabo's + GLideN64 each time.

**Comment by oddMLan** [comment]
> Hypothetically though, if GLideN64 fixes all the FB communications issues that these built-in cheats address, I would say just don't bother with the RDB, just make "Enable Game Fixes" only be effective when Jabo's 1.6/1.7 plugin is used regardless of the game or RDB

I agree with this.

**Comment by project64** [comment]
if I make the changes to the rdb, so I say which plugin it applies to then it will in essence be on just for jabo 1.6/1.7 plugin

**Comment by cxd4** [comment]
In some contingencies you may not have a choice I guess.  You could hypothetically be forced to put them in something like the RDB, so that the cheats will continue to hack in the fixes for something that even GLideN64 might still not have fixed yet.  The best forwards-compatibility with future plugins and their chance to emulate the thing right tho is just to prevent the cheat fixes from applying outside of just closed-source first-party plugins like Jabo's.  The RDB might or might not be necessary otherwise for other plugins like glide...but even then the cheat code could be manually added by the user.  Come to think of it, the CHT file would be an excellent choice compared to the RDB.  Why not just make the Project64.cht file contain the game fixes and enable them by default?  That way we are able to see them within the cheats GUI and see the exact code being applied.

**Comment by project64** [comment]
it use to be in the cheat file, but it requires the end user to know about it and then turn it on, 99% will not.

so better to have a system where it is turned on by default, and from an end user perspective it is not really a cheat, it is a hack. I doubt people are looking at the cheat. 

In essence should it be on by default and turned off for certain plugins, or off by default and turned on just for certain plugins, the latter is the one I am suggesting.

**Comment by cxd4** [comment]
I guess I just assumed you could say, if (!DisableGameFixes), check if gfx plugin starts with "Jabo's", if so turn on certain cheats from the CHT file.  But according to what you said you don't have that level of control over the CHT file for whatever reason.

**Comment by cxd4** [comment]
Here, example:
![chtexample](https://cloud.githubusercontent.com/assets/1396303/6049580/c21d9abc-ac86-11e4-86c5-9130e6e1c470.png)

**Comment by project64** [comment]
which is confusing to most users, they will not use it and there is nearly no point of having it.

I want project64 to work out of the box with out users having to modifying things in general

**Comment by cxd4** [comment]
When I say the checkbox defaults on, I mean that the cheat is used by default.

So it would work right out of the box without the user having to change anything.
The checkbox for that cheat code would only be turned off (and the cheat disabled) if plugin name doesn't start with "Jabo's".

**Comment by cxd4** [comment]
I'm talking about automation, not the user manually forcing the cheat on or off.

It's not too different from the automation you're considering to invest implementing to the RDB file.

**Comment by Fanatic-64** [comment]
Some cheats are used for issues not related to the plugins, for example Chamelon Twist has a cheat that allows it to work with the Controller Pak enabled, otherwise it requires disabling it in the controller settings to be able to play.

**Comment by project64** [comment]
ok made a change that sets the hack is only applied to jabo's plugin and pj64 glide version (https://github.com/project64/project64/commit/491377c2a4a68321023cbdf0a262f056afb7178c)

as discussed this is  a gfx plugin issue, there is nothing else I can do in the core.

So I think this can be closed now.

**Comment by Fanatic-64** [comment]
We should also activate that cheat for (death--droid's 1964 Aristotle's Mudlord's) Rice Video, Glide64, glN64, Direct64, TR64 and z64, since all those plugins also suffer the delay issue.

**Comment by project64** [comment]
true, but as they do not come with the emu I did not add them, I am guessing over 99% of people will either use jabo's or glide I do not think it is much of an issue.

**Comment by cxd4** [comment]
Encouraging plugins to continue to refuse to implement accurate frame buffer storage is NOT the correct way of looking at this Fanatic.  Of those mentioned, Direct64 is closed-source with its author being no longer of this world, but if it was really that big of a deal this would be moved to the CHT file and activated manually by the end user.

It really is not that hard to enforce frame buffer writes.  `glReadPixels` is just slow naturally, but that doesn't create an excuse for supporting third-party plugins which refuse to implement this PROPERLY, by adding in another dimension of RDB-ish hacking when we already have plugins which do properly support this.  I'm already reluctant enough to agree with it applying to just Jabo's plugin itself.

Now stop filing hack requests and design suggestions as GitHub issues, and focus on emulator itself.

**Comment by project64** [comment]
I was inclined to just put in for jabos plugin and not for pj64's glide64 since we got the source and can fix it, but if I did that I am sure people would complain that there was a regression in the plugin. So I guess it can stay that way to I get a real fix in for that one

**Comment by cxd4** [comment]
Yeah I would rather leave the hacks out of Glide64 as a constant reminder to have a look at implementing it.  I have an idea how to do it with basic acceleration, but it might not handle external things like depth buffer read/write.

In the worst case GLideN64 I'm sure would have it implemented.  The reason this issue was filed really just has nothing to do with what the discussion's moved on to and is a separate question.

**Comment by weinerschnitzel** [comment]
Maybe allow plugins to add permanent cheats?

Then a separate option to "Enable Legacy Game Hacks" that would be forced for plugins that predate that spec

Developer can move game hacks in question to plugin as discovered.

Enabling the cheat for specific plugins is complicated as another plugin could be created and if its name doesn't fit criteria, it may or may not work. Granted a person could add their plugin name to the source code to correctly use it, but that goes against the "out of box" approach some what. 

I think these issues should be handled between the plugin and emulator. How many of these hacks are for plugin issues?

**Comment by Frank-74** [comment]
@project64 The rdb plugin select hack option added with https://github.com/project64/project64/commit/491377c2a4a68321023cbdf0a262f056afb7178c, doesn't seem to be working with Pilot Wings.

**Comment by project64** [comment]
@Frank-74 in what way is it not working?

**Comment by Frank-74** [comment]
@project64 I believe the hack is supposed to be applied to Pilot Wings automatically if the plugin is Jabo's Direct3D8?.  But its not applied at all.  I still see the black square/land shadow thingy under the glider.
Also with the fix applied (I added it to the cheats to check), TLB must be off.  The fix breaks TLB somehow and gives you the funny squares thing after completing a mission.

**Comment by cxd4** [comment]
This would be so much easier to debug if we just had the cheats in the cheat file not RDB file. :P

Not really supposed to have game auto-fix cheats in the RDB file.  The CHT file could be given that power instead of the RDB file so that it would automatically turn the checkbox for the cheat on if you are using Jabo's plugin, off if you were using some other plugin, then the user could see the Cheats UI to see whether it's checked or unchecked automatically or even change it.  But putting cheats in the RDB like this just makes everything more confusing and hidden from everyone.

**Comment by project64** [comment]
the hack should only be applied to jabo's and glide64 now there is nothing to do with the core, I think this issue can be closed now.

**Comment by cxd4** [comment]
We know it can.  But @Fanatic-64 doesn't seem to know how to look for the "Close" button.

**Comment by Fanatic-64** [comment]
@zilmar, feel free to close issues that appear solved.

Though I may add, having tested with Jabo D3D8 1.7, without the cheat fix, and with Counter Factor=1, the delay opening the menu is only of about 1 second. So maybe we should add that setting to help other plugins.
